csrc/preseg_passes/move_repeat_forward.cpp

```diff
@@ -0,0 +1,480 @@
+    // dominator and see if that's a valid position
+    const auto post_dominators = getAllPostDominators(reshape_output_tv);
+
+    TensorView* target_tv = nullptr;
+    for (const auto& post_dominator : post_dominators | std::views::reverse) {
+      NVF_ERROR(
+          post_dominator->isA<TensorView>(),
+          "Expected to be a tensor: ",
+          post_dominator->toString());
+
+      if (auto it = repeat_id_map.find(post_dominator->as<TensorView>());
+          it != repeat_id_map.end() && it->second != nullptr) {
+        target_tv = post_dominator->as<TensorView>();
+        break;
+      }
+    }
+
+    // No valid position found
+    if (target_tv == nullptr) {
+      return std::nullopt;
+    }
+
+    return std::make_pair(target_tv, repeat_id_map);
+  }
+
+  // Move the repeat right after target_tv. All tensors between the
+  // reshape output tensor and the target tensor are updated by
+  // cancelling the repetition, i.e., shrinking the extent of the
+  // repeated ID back to the original extent.
+  //
+  // More concretely, consider a simple case like below (2x repetition):
+  //
+  // t0: [i0, b1(2)]
+  // t1 = reshape(t0); // [i0*2]
+  // t2 = op1(t1); // [i0*2]
+  // t3 = op2(t2); // [i0*2]
+  //
+  // Suppose we want to move the repeating reshape after t2. This
+  // pattern is going to be transformed to:
+  //
+  // t0: [i0, b1(2)]
+  // t4 = squeeze(t0, {1}); // [i0]
+  // t2 = op1(t4); // [i0]
+  // t5 = broadcast(t2); // [i0, b2(1)]
+  // t6 = expand(t5); // [i0, b2(2)]
+  // t7 = reshape(t6); // [i0*2]
+  // t3 = op2(t7); // [i0*2]
+  //
+  // Notice that op1 now operates on an ID with the pre-repeat
+  // extent.
+  //
+  // This transformation involves 1) squeezing the broadcast ID
+  // representing the repetition factor; 1) changing the extent of t2 from
+  // i0*2 to i0; 2) inserting the new repeat expr sequence after t2;
+  void moveTo(
+      const StaticRepeatInfo& info,
+      TensorView* target_tv,
+      const std::unordered_map<TensorView*, IterDomain*>& repeat_id_map) {
+    auto target_tv_repeat_id = repeat_id_map.at(target_tv);
+
+    auto target_logical =
+        TensorDomain::noReductions(target_tv->getLogicalDomain());
+    auto logical_it = std::ranges::find(target_logical, target_tv_repeat_id);
+    NVF_ERROR(logical_it != target_logical.end());
+    auto target_repeat_id_pos =
+        std::distance(target_logical.begin(), logical_it);
+
+    // Squeeze the broadcast of the tensor that was the input
+    // to the original repeat pattern
+    TensorView* squeeze_out = squeezeRepeatFactorBroadcast(info);
+
+    // Replace the repeated extent with the original
+    // pre-repeat extent
+    replaceRepeatedExtents(info, target_tv, repeat_id_map);
+
+    // Redirect the use of the original repeated tensor to the
+    // squeezed tensor
+    ir_utils::replaceValInAllExprInputsAndFusionOutputs(
+        info.reshape_output_tv, squeeze_out);
```

Comment:
Does this populate the change to downstream ops?

IIUC, we want to move this down stream to the farthest operation.

e.g.
```
  // t0 [i0, b1(2)]
  // t1 [i0*2]
  // t2 = op1(t1); // [i0*2]
  // t3 = op2(t2); // [i0*2]
  // t4 = op3(t3); // [i0*2]
```

If we need to move it to t4. IIUC, the replay here only changes the first operations after the reshape?
```
  // t0 [i0, b1(2)]
  // t1 [i0*2]
  // squeeze_out = squeeze(t0, {1}); // [i0]
  // t2 = op1(squeeze_out); // [i0]
  // t3 = op2(t2); // [i0*2]
  // t4 = op3(t3); // [i0*2]
```

Don't we need to update t2/t3 as well? Since we are changing the domain of inputs, it's not just a simple replace, we actually needed to replay the chain of operations.

Reply:
Did I miss something obvious here?
I do see that we have chains of ops in the test that should trigger this, but looks like tests are all cleared..

Reply:
If we did just `replaceValInAllExprInputsAndFusionOutputs`, the fusion would look like:

```
  // t0 [i0, b1(2)]
  // t1 [i0*2]
  // squeeze_out = squeeze(t0, {1}); // [i0]
  // t2 = op1(squeeze_out); // [i0*2]
  // t3 = op2(t2); // [i0*2]
  // t4 = op3(t3); // [i0*2]
```

Note that `t2`'s extent should have been `i0*2`. This is clearly inconsistent. However, since we do `replaceRepeatedExtents` before this point, the fusion should actually look like:

```
  // t0 [i0, b1(2)]
  // t1 [i0*2]
  // squeeze_out = squeeze(t0, {1}); // [i0]
  // t2 = op1(squeeze_out); // [i0]
  // t3 = op2(t2); // [i0]
  // t4 = op3(t3); // [i0*2]
```

Reply:
:face_with_head_bandage: Got'ya. sorry for the false alarm. Don't know how I missed that one.